keps/sig-cluster-lifecycle/0032-create-a-k8s-io-component-repo.md

```diff
@@ -0,0 +1,245 @@
+
+### Goals
+
+- Make it easy for a component to correctly adopt ComponentConfig.
+- Avoid moving code into _k8s.io/apiserver_ which does not strictly belong to an etcd-based, API group-serving apiserver. Corollary: remove etcd dependency from components.
+- Components should be consistent in how they load (and write) configuration, and merge config with CLI flags.
+- Factor out command- and flag-building code to a shared place.
+- Factor out common HTTPS endpoints describing a component's status.
+- Make the core Kubernetes components utilize these new packages.
+- Have good documentation about how to build a component with a similar look and feel as core Kubernetes components.
+- Increase test coverage for the configuration, command building, and HTTPS server areas of the component code.
+- Break out OpenAPI definitions and violations for the current ComponentConfigs from the core repo to a dedicated place per-component. With auto-generated OpenAPI specs for each component ComponentConfig consumers can validate/vet their configs without running the component.
+
+### Success metrics
+
+- All core Kubernetes components (kube-apiserver, kube-controller-manager, kube-scheduler, kubelet, kube-proxy, kubeadm) are using these shared packages in a consistent manner.
+- Cloud providers can be moved out of core without having to depend on the core repository.
+  - Related issue: [https://github.com/kubernetes/kubernetes/issues/69585](https://github.com/kubernetes/kubernetes/issues/69585)
+- It's easier for _kubeadm_ to move out of the core repo when these component-related packages are in a "public" staging repository.
+
+### Non-goals
+
+- Graduate any ComponentConfig API versions (in this proposal).
+- Make this library toolkit a "generic" cloud-native component builder. Such a toolbox, if ever created, could instead consume these packages. In other words, this repository is solely focused on Kubernetes' components' needs.
+- Fixing _all the problems_ our components have right now, and expanding this beyond what's really necessary. Instead working incrementally, and starting to break out some basic stuff we _know_ every component must handle (e.g. configuration and flag parsing)
+
+### Related proposals / references
```

Comment:
Some other relevant docs you might want to include:
- [Graduating KubeletFlags subfields to KubeletConfiguration](https://docs.google.com/document/d/18-MsChpTkrMGCSqAQN9QGgWuuFoK90SznBbwVkfZryo/edit#heading=h.ulrt23q582vb)
  - This one is about flag-merging trickiness.
- [Platform-Specific Component Configuration (DRAFT)](https://docs.google.com/document/d/1rfSq9PGn_b7ILvWXjkgU9R6h5C84ywbKUrtJWo0IrYw/edit#heading=h.iwuyjwpwphle)
  - This one is unfinished; my line of reasoning was too strict in the original and I haven't had time to come back to it, but we should think about how to handle platform-specific config.
- [Making /configz better](https://docs.google.com/document/d/1kNVSdw7H9EqyvI2BYH4EqwVtcg9bi9ZCPpJs8aKZFmM/edit#heading=h.a14lyh7vivpj)
  - This is an *OLD* doc I wrote with some ideas on how to make /configz better.

Reply:
Thanks for these, absolutely, will do!